server/routing/pubsub/manager.go

```diff
@@ -0,0 +1,282 @@
+
+	return nil
+}
+
+// SetOnLabelAnnouncement sets the callback for received label announcements.
+// This callback is invoked for each valid announcement received from remote peers.
+//
+// The callback should:
+//   - Convert wire format ([]string) to labels.Label
+//   - Build enhanced keys using BuildEnhancedLabelKey()
+//   - Store labels.LabelMetadata in datastore
+//
+// Example:
+//
+//	manager.SetOnLabelAnnouncement(func(ctx context.Context, ann *LabelAnnouncement) {
+//	    for _, labelStr := range ann.Labels {
+//	        label := labels.Label(labelStr)
+//	        key := BuildEnhancedLabelKey(label, ann.CID, ann.PeerID)
+//	        // ... store in datastore ...
+//	    }
+//	})
+func (m *Manager) SetOnLabelAnnouncement(fn func(context.Context, *LabelAnnouncement)) {
+	m.onLabelAnnouncement = fn
+}
+
+// handleMessages is the main message processing loop.
+// It runs in a goroutine and processes all incoming label announcements.
+//
+// Flow:
+//  1. Wait for next message from subscription
+//  2. Skip own messages (already cached locally)
+//  3. Unmarshal and validate announcement
+//  4. Invoke callback for processing
+//
+// Error handling:
+//   - Context cancellation: Normal shutdown, exit loop
+//   - Invalid messages: Log warning, continue processing
+//   - Unmarshal errors: Log warning, continue processing
+//
+// This goroutine runs for the lifetime of the Manager.
+func (m *Manager) handleMessages() {
+	for {
+		msg, err := m.sub.Next(m.ctx)
+		if err != nil {
+			// Check if context was cancelled (normal shutdown)
+			if m.ctx.Err() != nil {
+				logger.Debug("Message handler stopping", "reason", "context_cancelled")
+
+				return
+			}
+
+			// Log error but continue processing
+			logger.Error("Error reading from labels topic", "error", err)
+
+			continue
+		}
+
+		// Skip our own messages (we already cached labels locally)
+		if msg.ReceivedFrom == m.host.ID() {
+			continue
+		}
+
+		// Parse and validate announcement
+		announcement, err := UnmarshalLabelAnnouncement(msg.Data)
+		if err != nil {
+			logger.Warn("Received invalid label announcement",
+				"from", msg.ReceivedFrom,
+				"error", err,
+				"size", len(msg.Data))
+
+			continue
+		}
+
+		logger.Debug("Received label announcement",
+			"from", msg.ReceivedFrom,
+			"cid", announcement.CID,
+			"peer", announcement.PeerID,
+			"labels", len(announcement.Labels))
+
+		// Invoke callback for processing
+		if m.onLabelAnnouncement != nil {
+			// Use context from Manager, not from message
+			m.onLabelAnnouncement(m.ctx, announcement)
+		}
+	}
+}
+
+// GetTopicPeers returns the list of peers subscribed to the labels topic.
+// This is useful for monitoring network connectivity and debugging.
+//
+// Returns:
+//   - []string: List of peer IDs (as strings)
+func (m *Manager) GetTopicPeers() []string {
+	peers := m.topic.ListPeers()
+	peerIDs := make([]string, len(peers))
+
+	for i, p := range peers {
+		peerIDs[i] = p.String()
+	}
+
+	return peerIDs
+}
```

Comment:
i do not know if it is efficient to have only one topic across the network. i think for the time being, its okay, but later on, it would be useful to subscribe to a set of topics regarding the publication data (e.g. specific labels -- skills, domains, etc) which can be policy-driven via config in the server.
note that in this case, this method will return all the peers for the topic across the whole network